lib/livebook/hubs.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Livebook.Hubs do
+  @moduledoc false
+
+  alias Livebook.Storage
+  alias Livebook.Hubs.{Fly, Hub, HubProvider}
+
+  defmodule NotFoundError do
+    @moduledoc false
+
+    defexception [:id, :type, plug_status: 404]
+
+    def message(%{id: id, type: "fly"}) do
+      "could not find a fly entry matching #{inspect(id)}"
+    end
+
+    def message(%{id: "fly-" <> id}) do
+      message(%{id: id, type: "fly"})
+    end
+  end
+
+  @namespace :hubs
+
+  @doc """
+  Gets a list of hubs from storage.
+  """
+  @spec fetch_hubs() :: list(Hub.t())
+  def fetch_hubs do
+    for fields <- Storage.current().all(@namespace) do
+      fields
+      |> to_struct()
+      |> HubProvider.to_hub()
+    end
+  end
+
+  @doc """
+  Gets one hub from storage.
+
+  Raises `NotFoundError` if the hub does not exist.
+  """
+  @spec fetch_fly!(String.t()) :: Fly.t()
+  def fetch_fly!("fly-" <> id = full_id) do
+    case Storage.current().fetch(@namespace, full_id) do
+      :error -> raise NotFoundError, id: id, type: "fly"
+      {:ok, fields} -> to_struct(fields)
+    end
+  end
+
+  @doc """
+  Gets one provider from storage.
+
+  Raises `NotFoundError` if does not exist.
+  """
+  @spec provider_by_id!(String.t()) :: String.t()
+  def provider_by_id!(id) do
+    case Storage.current().fetch(@namespace, id) do
+      :error -> raise NotFoundError, id: id
+      {:ok, _} -> id_to_provider(id)
+    end
+  end
+
+  defp id_to_provider("fly-" <> _), do: "fly"
+
+  @doc """
+  Checks if Fly already exists.
+  """
+  @spec fly_exists?(Fly.Organization.t()) :: boolean()
+  def fly_exists?(%Fly.Organization{id: id}) do
+    case Storage.current().fetch(@namespace, "fly-" <> id) do
+      :error -> false
+      {:ok, _} -> true
+    end
+  end
+
+  @doc """
+  Saves a new hub to the configured ones.
+  """
+  @spec save_fly(Fly.t()) :: Fly.t()
+  def save_fly(%Fly{organization: organization} = fly) do
+    id = "fly-" <> fly.id
+
+    attributes =
+      fly
+      |> Map.delete(:id)
+      |> Map.replace!(:organization, to_ets(organization))
+      |> to_ets()
```

Comment:
```suggestion
  def save_hub(struct) do
    map =
      struct
      |> Map.from_struct()
      |> Map.update!(:id, &provider_type(struct) <> "-" <> id)
```

Reply:
instead of `Hub.save_fly` you will do `Fly.insert_hub` and `Fly.insert_hub` calls `Hub.insert_hub`.